pkg/operator/operator.go

```diff
@@ -241,76 +273,237 @@ func (k *Kontroller) withLeaderElection() error {
 	if len(rebootingNodes) >= maxRebootingNodes {
-		glog.Infof("Found %d (of max %d) rebooting nodes; waiting for completion", len(rebootingNodes), maxRebootingNodes)
 		for _, n := range rebootingNodes {
 			glog.Infof("Found node %q still rebooting, waiting", n.Name)
 		}
-		return
+		glog.Infof("Found %d (of max %d) rebooting nodes; waiting for completion", len(rebootingNodes), maxRebootingNodes)
+		return nil
 	}
 
+	// find nodes which want to reboot
 	rebootableNodes := k8sutil.FilterNodesByAnnotation(nodelist.Items, wantsRebootSelector)
+	rebootableNodes = k8sutil.FilterNodesByRequirement(rebootableNodes, notBeforeRebootReq)
+
+	// Don't even bother if rebootableNodes is empty. We wouldn't do anything anyway.
 	if len(rebootableNodes) == 0 {
-		return
+		return nil
 	}
 
+	// find the number of nodes we can tell to reboot
 	remainingRebootableCount := maxRebootingNodes - len(rebootingNodes)
-	// pick N of the eligible candidates to reboot
+
+	// choose some number of nodes
 	chosenNodes := make([]*v1api.Node, 0, remainingRebootableCount)
 	for i := 0; i < remainingRebootableCount && i < len(rebootableNodes); i++ {
 		chosenNodes = append(chosenNodes, &rebootableNodes[i])
 	}
 
+	// set before=reboot=true for the chosen nodes
 	glog.Infof("Found %d nodes that need a reboot", len(chosenNodes))
-	for _, node := range chosenNodes {
-		k.markNodeRebootable(node)
+	for _, n := range chosenNodes {
+		err = k.mark(n.Name, constants.LabelBeforeReboot, k.beforeRebootAnnotations)
+		if err != nil {
+			return fmt.Errorf("Failed to mark node for before reboot checks: %v", err)
+		}
+		glog.Infof("Waiting for configured before-reboot annotations on node %q", n.Name)
+		glog.V(4).Infof("Waiting for the following annotations: %v", k.beforeRebootAnnotations)
+	}
+
+	return nil
+}
+
+// markAfterReboot gets nodes which have completed rebooting and marks them with
+// the after-reboot=true label. A node with the after-reboot=true label is still
+// considered to be rebooting from the perspective of the update-operator, even
+// though it has completed rebooting from the machines perspective.
+// If there is an error getting the list of nodes or updating any of them, an
+// error is immediately returned.
+func (k *Kontroller) markAfterReboot() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed listing nodes: %v", err)
+	}
+
+	// find nodes which just rebooted
+	justRebootedNodes := k8sutil.FilterNodesByAnnotation(nodelist.Items, justRebootedSelector)
+	// also filter out any nodes that are already labeled with after-reboot=true
+	justRebootedNodes = k8sutil.FilterNodesByRequirement(justRebootedNodes, notAfterRebootReq)
+
+	if len(justRebootedNodes) > 0 {
+		glog.Infof("Found %d rebooted nodes, setting label %q to %q", len(justRebootedNodes), constants.LabelAfterReboot, constants.True)
 	}
+
+	// for all the nodes which just rebooted, remove any old annotations and add the after-reboot=true label
```

Comment:
by any old annotations, you mean after reboot annotations from a prior post reboot window?